local.go

```diff
@@ -124,37 +134,49 @@ func Local(p Problem, initX []float64, settings *Settings, method Method) (*Resu
 	}, err
 }
 
-func minimize(settings *Settings, method Method, p *Problem, stats *Stats, optLoc *Location, startTime time.Time) (status Status, err error) {
+func minimize(p *Problem, method Method, settings *Settings, stats *Stats, optLoc *Location, startTime time.Time) (status Status, err error) {
 	loc := &Location{}
 	copyLocation(loc, optLoc)
-	xNext := make([]float64, len(loc.X))
 
 	methodStatus, methodIsStatuser := method.(Statuser)
 
-	evalType, iterType, err := method.Init(loc, xNext)
+	var op Operation
+	op, err = method.Init(loc)
 	if err != nil {
-		return Failure, err
+		status = Failure
+		return
 	}
 
 	for {
-		if p.Status != nil {
-			// Check the function status before evaluating.
-			status, err = p.Status()
-			if err != nil || status != NotTerminated {
-				return
+		switch op {
+		case NoOperation:
+
+		case InitIteration:
+			panic("optimize: Method returned InitIteration")
+
+		case PostIteration:
+			panic("optimize: Method returned PostIteration")
+
+		case MajorIteration:
+			stats.MajorIterations++
+			if loc.F <= optLoc.F {
+				copyLocation(optLoc, loc)
 			}
-		}
+			status = checkConvergence(optLoc, settings)
 
-		// Perform evalType evaluation of the function at xNext and store the
-		// result in location.
-		evaluate(p, evalType, xNext, loc, stats)
-		// Update the stats and optLoc.
-		update(loc, optLoc, stats, iterType, startTime)
-		// Get the convergence status before recording the new location.
-		status = checkConvergence(optLoc, iterType, stats, settings)
+		default:
+			if p.Status != nil {
+				status, err = p.Status()
+				if err != nil || status != NotTerminated {
+					return
+				}
+			}
+			evaluate(p, loc, op, stats)
+		}
 
 		if settings.Recorder != nil {
-			err = settings.Recorder.Record(loc, evalType, iterType, stats)
+			stats.Runtime = time.Since(startTime)
```

Comment:
I wonder if we shouldn't update Runtime only once per iteration so that the Recorder and checkLimits() get the same value. The complete opposite would be replacing Runtime with StartTime so that the Recorder and checkLimits() are forced to compute the actual runtime at the moment when they need it.